### ISSUE TITLE
Garage door switch type

### ISSUE THREAD

**Issue by Patrik-svobodik** [howanghk/homebridge-ewelink#23]
Woudl it be possible to change the type of the switch to garage door? I am using the ewelink connected garage door opener (don’t remember the name) and it’s working great, but I would like to have it show up as a garage door rather than a switch

**Comment by howanghk** [comment]
Hi @Patrik-svobodik, in HomeKit a GarageDoorOpener is quite different from a Switch. A switch only have a on/off toggle. A GarageDoorOpener have a CurrentDoorState, TargetDoorState and ObstructionDetected characteristics, which I am not sure if your ewelink garage door opener support them or not. Even if it does support them, as I don't personally own such device, it will be hard or impossible for me to implement it.

**Comment by Patrik-svobodik** [comment]
The device supports all info that homekit needs, I researched that (the target door state can be assumed and ObstructionDetected ignored as my garage door detects that automatically), if you want, I can give you as much diagnostic data as you want, it operates as a switch that is on when it is opened and off when closed and on turning on/off changes state (tries to open/close)

**Comment by S474N** [comment]
+1

All states are user defined. There is no problem :)

**Comment by Patrik-svobodik** [comment]
And one more detail, if it tries to change state (opened/closed) but the sensor doesn’t trigger it automatically overrides it to the real state

**Comment by S474N** [comment]
Which device do you have as garage door opener?

**Comment by Patrik-svobodik** [comment]
I use this as my garage door opener:
https://www.google.com/amp/s/m.aliexpress.com/item/32889505304.html

st 27. 3. 2019 v 23:10 odesílatel S474N <notifications@github.com> napsal:

**Comment by S474N** [comment]
I bought this one, it's more cheaper:
https://www.aliexpress.com/item/SONOFF-5V-12V-220V-RF-433MHz-Wifi-Switch-Wireless-Relay-Module-Remote-Switch-for-Android-IOS/32978091649.html

**Comment by Bogdis** [comment]
@S474N can you see if your garage door is open or closed with that relay? I have almost the same one but cant se if it’s open or not so i have set up a camera for that purpose, but I would like to change to the one Patrik linked to but i cant find any good info if it works properly with homebridge and shows the right state and can sent notifications if i forget to close the garage door.

**Comment by Patrik-svobodik** [comment]
@Bogdis the one I linked works perfectly and detects the state, only downside is that it shows up as a switch instead of a garage door opener. It works reliably for me

**Comment by Bogdis** [comment]
@Patrik-svobotic so you mean that you can see in home app that the switch is on when gragage door is open and off when closed? Do you get notifications if you forget the garage open through home app? I saw that there is a version for tuya to and with the tuya-lan plugin you can get it to show upp as a garage door but there are still som problems there too at least for this user, 

https://github.com/AMoo-Miki/homebridge-tuya-lan/issues/7

Could you test if you’re device can work with tuya app or smartlife app and tuya-lan plugin?

**Comment by Patrik-svobodik** [comment]
I don’t know anything about tuya and I don’t plan on trying it, sorry

**Comment by Bogdis** [comment]
@Patrik-svobodik ok no problem 😁 what about the other things I asked about?

**Comment by Patrik-svobodik** [comment]
The switch thing is right, you don’t get notifications saddly

čt 28. 3. 2019 v 22:02 odesílatel Bogdis <notifications@github.com> napsal:

**Comment by S474N** [comment]
I used wired magnetic sensor to my smart system and up Aqara Door Sensor (via Zigbee do Aqara Hub).

**Comment by howanghk** [comment]
@S474N, if you are using Aqara Hub, you should look into [homebridge-mi-aqara](https://github.com/YinHangCode/homebridge-mi-aqara) instead, which is more robust and support many type of accessories. Or if you get the latest Aqara Hub, it support HomeKit natively. You should not stick to sonoff / ewelink as they don't bother to support HomeKit, and there is no hobbyist developer adding support for it.

@Patrik-svobodik, I understand you really want it to be implemented. However, I don't have the incentive to work on it, as 1. I don't use the device; 2. I don't get paid to work for ewelink.

Therefore, you are left with 3 choices:

1. Implement the functionality yourself (**best option**)
This project is open source and have a relaxed license that you are free to modify the code and share your work. You can look into the source code, it's written in Javascript which is fairly easy to understand and write. Once you're done with the extra functions, you are welcome to send a pull request so the functions can be merged into this fork. This way you get what you want, and benefits all users around the world.

2. Hire a developer to work on the Homebridge ewelink plugin
This will probably be the most expensive option. It will cost $300+ USD for the work, and you will have to provide the device to the developer you hire.

3. Choose another vendor (**easiest**)
There are many other vendors offering home automation devices that come with native HomeKit support or have a decent Homebridge plugin. You may want to look into them.
Apple have a list of supported HomeKit garage door on their website: https://www.apple.com/ios/home/accessories/#section-garage

**Comment by Bogdis** [comment]
@S474N how did you wired the wired magnetic sensor to the relay?

**Comment by Patrik-svobodik** [comment]
@howanghk I will try, do you have any docs for homekit?

**Comment by Patrik-svobodik** [comment]
Ok, so I looked through your code and it is pretty easy to understand, but I ran into one problem, uiid of the device is 1 and I have no other way to identificate it

**Comment by howanghk** [comment]
@Patrik-svobodik uiid 1 = SOCKET, so it's just a plain simple socket with basic on / off toggle only. 🤔

**Comment by Patrik-svobodik** [comment]
I know that, I read the code, wondering how could I recognise the device

pá 29. 3. 2019 v 17:34 odesílatel Wong Ho Wang <notifications@github.com>
napsal:

**Comment by howanghk** [comment]
For documentation you may want to look at the [Writing Plugins](https://github.com/nfarina/homebridge#writing-plugins) section in the homebridge README.

**Comment by howanghk** [comment]
Right, there might be no way to recognize it. You may add new config value to indicate a specific device should be changed to garage door type.

**Comment by S474N** [comment]
howanghk: this plugin is only for Mijia Gateway (2.gen), not for Aqara Gateway (3.gen with HK support). I have Aqara, and this cannot be added to NodeRED or another systems, because there is no developer mode (as on Mijia). 

bogdis: I have smart system and magnet is in connected to this system. 

patrik-svobodnik: this is better web and here you are GarageDoorOpener service:
http://auto.caitken.com/posts/2018/09/09/nodered-homekit-reference#ServiceGarageDoorOpener
